x-pack/plugins/fleet/server/plugin.ts

```diff
@@ -382,4 +394,39 @@ export class FleetPlugin
     licenseService.stop();
     this.telemetryEventsSender.stop();
   }
+
+  private getAgentServiceFactory(internalEsClient: ElasticsearchClient): AgentServiceFactory {
+    type ESFunc = (client: ElasticsearchClient, ...rest: any[]) => any;
+    type ESFuncArgs<T extends ESFunc> = T extends (client: any, ...rest: infer R) => any
+      ? R
+      : never;
+
+    return {
+      asScoped: (req) => {
+        const useInternalClientIfSuperuser =
+          <TFunc extends ESFunc>(func: TFunc) =>
+          (...rest: ESFuncArgs<TFunc>) => {
+            if (!checkSuperuser(req)) {
+              throw new Error(`User does not have adequate permissions to access Fleet agents.`);
+            }
```

Comment:
When we support non-superusers, we'd replace this check with a Kibana privileges check.